IS-312: Inject User Data into GrowthBook for Segmentation

harishv7 opened the pull request:
## Problem

We need to have context of user attributes for us to do targeted rollouts/experiments.

Closes IS-312

## Solution

We inject information at both the authentication and authorization middlewares based on what we know about the user at that point in time.

Note: To be reviewed in conjunction with IS-310 which introduces GrowthBook for BE.

**Breaking Changes**

<!-- Does this PR contain any backward incompatible changes? If so, what are they and should there be special considerations for release? -->

- [ ] Yes - this PR contains breaking changes
  - Details ...
- [X] No - this PR is backwards compatible

harishv7 commented:
Note: Tests will pass once #928 is merged into this

QiluXie commented on src/types/express/request.d.ts:
```diff
@@ -0,0 +1,14 @@
+
+import { FeatureFlags } from "@root/types/featureFlags"
+
+// to make the file a module and avoid the TypeScript error
```
nit: added more details in comment for the nature of the error we are trying to avoid here 

```suggestion
// Empty export to make this file a module therefore allow global augmentation (avoiding the related error)
```

harishv7 replied:
updatedd

harishv7 commented on src/types/featureFlags.ts:
```diff
@@ -3,3 +3,12 @@
+  email: string
+  githubId?: string
+  siteName?: string
+  role?: "email" | "admin"
```
Actually this is not needed now, but only if we need to segment by roles. I will remove this for now and we can evaluate if there is an actual need for it later

QiluXie reviewed (approved):
lgtm. Thanks @harishv7 for addressing the comments quickly.
Should merge #928 first and make sure the CI testing passed before merging this one.

harishv7 commented:
> lgtm. Thanks @harishv7 for addressing the comments quickly. Should merge #928 first and make sure the CI testing passed before merging this one.

@QiluXie all checks have passed. merging now, thanks
